### ISSUE TITLE
Build instructions for Windows

### ISSUE THREAD

**Issue by st2430** [OlivierJG/botansqlite3#2]
The build instructions provided with this package are not helpful when trying to build under Windows using MinGW or the Windows SDK. In particular, the `pkg-config --cflags botan-1.10` requires access to pkg-config, which is yet another obstacle to overcome using MinGW. While the readme.txt file is re commendably short and concise it isn't of much help to those of us using the Windows platform. It would be great if either a make file suitable for nmake was provided or at least a 'compile.bat' file that would allow a smooth compile and link of the botansqlite3 components. Many thanks for listening.
Staffan

**Comment by OlivierJG** [comment]
I agree with the need for this. Unfortunately I don't use Windows or have easy access to a Windows development machine, so all I can say is I'd welcome any improvements from someone who does.

For your assistance however, the pkg-config line simply returns the include path for botan for GCC's benefit (for example "-I/usr/include/botan-1.10"). It can simply be replaced with the appropriate botan include path for Windows.

Feel free to report back with instructions or a .bat file for Windows if you get it working.

**Comment by shunruo** [comment]
I use this on Windows without MinGW. And I think it's easy if you already got botan build, the botan readme file is detailed.

**Comment by st2430** [comment]
Thank you both for such a quick feedback. First of all I need to inform you that I'm not a C/C++ developer and am therefore struggling with anything that has to do with a C/C++ compiler. Saying that I did actually manage to perform a make of botan using MinGW (I think it was), but I never got the make install to work. But the make gave me the libs that I believe are necessary for the botansqlite3 install. Now that I have two specialists on the hook ;) I send you the compile.bat that I have created so far which doesn't work and maybe you can provide a corrected version that would work for me. sqlite3.c and codec.cpp compile OK but the compile/link of test_sqlite.cpp fails with no solution in sight. For your information I've got MinGW64 and Windows SDK 7.1 installed under Windows 7 (64-bit). Your help is highly appreciated.
Staffan

echo --- Compile BotanSQLite3
gcc -DSQLITE_HAS_CODEC -DSQLITE_THREADSAFE=1 -DSQLITE_ENABLE_FTS4 -DSQLITE_ENABLE_RTREE -s -O2 -o botansqlite3.o -c sqlite3.c
echo --- Compile codec.cpp
gcc -I . -msse2 -o codec.o -c codec.cpp 
ar rcs libbotansqlite3.a botansqlite3.o codec.o
gcc -I. -I./botan test_sqlite.cpp -o test_sqlite.exe  -L. -L./botan -lbotansqlite3 ./botan/botan.lib ./botan/botan.exp

**Comment by OlivierJG** [comment]
I can't tell you exactly what to do on Windows, for previously mentioned
reasons, but you should probably show the exact errors you run into. If it
fails in the linking stage as you indicated, you can probably google select
parts of the error to find the library you need.
On linux, the libraries required by botan (as listed by botan-config, an
executable which you should have on Windows also) are:
"-lbotan-1.10 -ldl -lpthread -lrt"
Those will certainly be different on Windows, but that should point you in
the right direction.

-Olivier JG

**Comment by st2430** [comment]
I made some mods to my compile.bat so it now looks like this:

gcc -DSQLITE_HAS_CODEC -DSQLITE_THREADSAFE=1 -DSQLITE_ENABLE_FTS4 -DSQLITE_ENABLE_RTREE -s -O2 -o botansqlite3.o -c sqlite3.c
gcc -I . -msse2 -o codec.o -c codec.cpp 
ar rcs libbotansqlite3.a botansqlite3.o codec.o
gcc -I. -I./botan test_sqlite.cpp -o test_sqlite.exe  -L. -L./botan -lbotansqlite3 -lbotan-1.10 -lpthread 

If I use the -ldl -lrt flags I get the message -ldl / -lrt not found, I don't know what that indicates.

BTW, I've got libbotan-1.10.a and libbotan-1.10.dll in the current directory.

The last compile generates lots of messages similar to this:

C:\Users\Staffan\AppData\Local\Temp\ccOu8ymr.o:test_sqlite.cpp:(.text+0x166): undefined reference to `__gxx_personality_sj0'
./libbotansqlite3.a(codec.o):codec.cpp:(.text+0x76): undefined reference to`__gxx_personality_sj0'
./libbotansqlite3.a(codec.o):codec.cpp:(.text+0x1d3): undefined reference to `__gxx_personality_sj0'
./libbotansqlite3.a(codec.o):codec.cpp:(.text+0x326): undefined reference to`__gxx_personality_sj0'

From searching the web I believe they are caused by a mixture of C and C++ code, whether it's true and how to fix I haven't a clue :(

**Comment by OlivierJG** [comment]
As I mentioned "Those will certainly be different on Windows, but that
should point you in the right direction", "ldl" "lrt" et all are what are
used on Linux and were just an example of the output of botan-config on
Linux.

A mixture of C and C++ code is exactly what botansqlite3 is... where that
becomes a problem I'm not sure. If you haven't already see:
http://stackoverflow.com/questions/7751640/undefined-reference-to-gxx-personality-sj0
http://stackoverflow.com/questions/2189681/undefined-references-when-trying-to-link-qt-app-with-my-static-library

... did you patch sqlite.c?

-Olivier JG

**Comment by st2430** [comment]
Yes, sqlite.c is changed with #define and #include added. I'd seen the links you posted before and tried by changing gcc to g++ for the test_sqlite.cpp compile but ended up with what I thought were the same messages. But redoing it now I realise that the list of error messsages is shorter and now starts with the following entries:

./libbotansqlite3.a(codec.o):codec.cpp:(.text+0xc55): undefined reference to `__imp__ZN5Botan4PipeC1EPNS_6FilterES2_S2_S2_'
./libbotansqlite3.a(codec.o):codec.cpp:(.text+0xc8f): undefined reference to`__imp__ZN5Botan4PipeC1EPNS_6FilterES2_S2_S2_'
./libbotansqlite3.a(codec.o):codec.cpp:(.text+0xcc9): undefined reference to `__imp__ZN5Botan4PipeC1EPNS_6FilterES2_S2_S2_'
./libbotansqlite3.a(codec.o):codec.cpp:(.text+0xe5d): undefined reference to`__imp__ZN5Botan4PipeD1Ev'
./libbotansqlite3.a(codec.o):codec.cpp:(.text+0xe8b): undefined reference to `__imp__ZN5Botan4PipeD1Ev'
./libbotansqlite3.a(codec.o):codec.cpp:(.text+0xeb9): undefined reference to`__imp__ZN5Botan4PipeD1Ev'

This looks botan-specific, I guess caused by a missing library of some sort?

Staffan

**Comment by st2430** [comment]
As I mentioned I did a make, which created libbotan-1.10.a, libbotan-1.10.dll, libbotan-1.10.dll.0 and libbotan-1.10.dll.0.3. The first two I copied into the same directory as botansqlite3 where I perform the compile. The dll's that end .0 and .0.3 I don't know the significance of and it makes no difference if they are there or not.

I didn't manage to get the make install working but does that create additional libraries? I thought it just copies the make result to the system paths...

Staffan

**Comment by st2430** [comment]
Shunruo, if you are still listening, you say you are using this in Windows, would you mind sharing the secret how you do it?

Staffan

**Comment by st2430** [comment]
Thanks to invaluable offline help from shunruo I've now finally managed to compile and successfully run test_sqlite in my Windows 7 environment. shunruo provided me with two make files among other things that successfully created the final result. While working on this I also realised that the patch to sqlite3.c as provided in botansqlite3 is not really required, instead the following little snippet can be used as input to the compiler in place of a modified version of sqlite3.c (I'm not in favour of modifying source code to make it fit ones own requirements):

#define SQLITE_HAS_CODEC 1
#include "sqlite3.c"
#include "codecext.c"

So the major steps to get from A to B using MinGW under Windows can be summarised as follows (this assumes that you have MinGW (or MinGW64) and python installed and that extension .py initiates python):
1. Download and unzip botan 1.10.3 into your MSYS home directory
2. Using MSYS:
   configure.py --cc=gcc --os=mingw --with-tr1=none --cpu=x86_32  (or x86_64)
   make
   make check
3. Run check.exe created by 'make check' to verify that botan works. ('make install' is not required and doesn't work either without substantial changes to the make file due to bugs in configure.py)
4. Download and unzip botanSQLite3 into your MSYS home directory
5. Copy the INCLUDE folder from botan-10.3\build to the botanSQLite3 folder. (You can delete the INTERNAL folder once copied to save space)
6. Copy botan-1.10\libbotan-1.10.a to the botanSQLite3 folder
7. Download the latest SQLite3 amalgamation source and save in the botanSQLite3 folder
8. Copy sqlite3.h and sqlite3ext.h to the botanSQLite3 INCLUDE folder
9. Create sqlite3mod.c using the source statements above.
10. Change sqlite3.c to sqlite3mod.c in mysqlite.makefile (provided by shunruo)
11. Using MSYS from the botanSQLite3 folder, run make
12. Run test_sqlite.exe in the resulting gccRelease folder to verify the installation.

I hope the above is somewhat correct. Now is the question how much of this can be added as a standard feature to botanSQLite3, but that's not for me to decide. :)

Thanks for all input that I've received chasing a solution to this.
Staffan

**Comment by st2430** [comment]
I've now run into another problem. I'm trying to compile the sqlite3mod.c module as described earlier using the Windows SDK and the following command:

nmake /f makefile.win

makefile.win contains instructions to compile sqlite3mod.c but it generates the following error messages:

C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(181) : error C2275: 'Pgno' : illegal use of this type as an expression
        c:\mingw64\msys\home\staffan\oosqlite\src\sqlite\sqlite3.c(8944) : see declaration of 'Pgno'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(181) : error C2146: syntax error : missing ';' before identifier 'nPage'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(181) : error C2065: 'nPage' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(183) : error C2275: 'Pgno' : illegal use of this type as an expression
        c:\mingw64\msys\home\staffan\oosqlite\src\sqlite\sqlite3.c(8944) : see declaration of 'Pgno'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(183) : error C2146: syntax error : missing ';' before identifier 'nSkip'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(183) : error C2065: 'nSkip' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(184) : error C2275: 'DbPage' : illegal use of this type as an expression
        c:\mingw64\msys\home\staffan\oosqlite\src\sqlite\sqlite3.c(8954) : see declaration of 'DbPage'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(184) : error C2065: 'pPage' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(186) : error C2275: 'Pgno' : illegal use of this type as an expression
        c:\mingw64\msys\home\staffan\oosqlite\src\sqlite\sqlite3.c(8944) : see declaration of 'Pgno'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(186) : error C2146: syntax error : missing ';' before identifier 'n'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(186) : error C2065: 'n' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(187) : error C2065: 'n' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(187) : error C2065: 'n' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(187) : error C2065: 'nPage' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(187) : error C2065: 'n' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(189) : error C2065: 'n' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(189) : error C2065: 'nSkip' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(192) : error C2065: 'n' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(192) : error C2065: 'pPage' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(192) : warning C4047: 'function' : 'DbPage **' differs in levels of indirection from 'int *'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(192) : warning C4024: 'sqlite3PagerAcquire' : different types for formal and actual parameter 3
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(196) : error C2065: 'pPage' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(196) : warning C4047: 'function' : 'DbPage *' differs in levels of indirection from 'int'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(196) : warning C4024: 'sqlite3PagerWrite' : different types for formal and actual parameter 1
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(197) : error C2065: 'pPage' : undeclared identifier
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(197) : warning C4047: 'function' : 'DbPage *' differs in levels of indirection from 'int'
C:\MinGW64\msys\home\Staffan\botanSQLite3\codecext.c(197) : warning C4024: 'sqlite3PagerUnref' : different types for formal and actual parameter 1
NMAKE : fatal error U1077: '"c:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\Bin\cl.EXE"' : return code '0x2'

Apparently the Windows compiler doesn't like the way codecext.c has been coded. Can I have some help resolving this please? A google search doesn't give anything of value. Many thanks.
Staffan

**Comment by shunruo** [comment]
That's why anothor patch exist, apply it  :D

**Comment by st2430** [comment]
OK, I must have misunderstood the format in codecext.c.patch. I thought that > means insert and < means remove, is it the opposite? I've never seen this patch format before...

**Comment by st2430** [comment]
Hmm, running a compare between your version and the original botanSQLite3 version seems to indicate that Olivier has implemented the patch in the base code. I'm using the original code as input to the compile. The difference here is that I'm using te Windows SDK compiler and nmake instead of the make under MSYS. Maybe without the patch it will work there? I'll give it a go just in case :)

**Comment by st2430** [comment]
Interesting, using Windows SDK the (old) version of codecext.c that you shunruo supplied (without the patch) compiles OK, the (patched) version supplied in Olivier's botanSQLite3 package does not compile correctly with the SDK. So there seems to be a compatibility issue here.

**Comment by shunruo** [comment]
You are still lost in bushes... My version is a newer one. The download version is OK for GNU make. but for nmake, you need to change it to my version(use the patch file to the download one).

**Comment by NathanaelA** [comment]
Any chance this patch and makefile are available somewhere?

**Comment by Feonik** [comment]
> Change sqlite3.c to sqlite3mod.c in mysqlite.makefile (provided by shunruo)

Where can I download this mysqlite.makefile?

**Comment by Feonik** [comment]
Sorry for my English. This is my instruction for MinGW:
- 1. Setup MSYS (http://www.mingw.org/wiki/MSYS)
- 2. Setup MinGW-w64 (http://mingw-w64.sourceforge.net) (i686-4.8.2-release-win32-sjlj-rt_v3-rev0.7z)
- 3. Set Path for MSYS and MinGW-w64 (...\mingw-w64-32-win32-4.8.2\bin;...\msys\1.0\bin;%path%)
- 4. Setup python-2.7.5.msi (http://www.python.org)
- 5. See instruction "Setup Python 2.7.2 and MinGW on Windows XP" + Comments (http://www.secker.nl/2012/03/08/setup-python-2-7-2-and-mingw-on-windows-xp/)
- 6. Unpack files

Botan-1.10.6.tbz
botansqlite3-master.zip
sqlite-amalgamation-3071000.zip

(I use sqlite-3.7.10 and change "codecext.c" for this. For sqlite-3.7.12.1 work without this.)

to directory
msys\1.0\home\Feonik\Botan-1.10.6\

List of ./Botan-1.10.6 like this:

/checks
/doc
/src
botan_version.py
codec.cpp
codec.h
codec_c_interface.h
codecext.c
configure.py
readme.txt
shell.c
sqlite3.c
sqlite3.h
sqlite3ext.h
sqlite3-amalgamation.patch
test_sqlite.cpp
- 7. In ./Botan-1.10.6 create file

== sqlite3mod.c ==

#define SQLITE_HAS_CODEC 1
#include "sqlite3.c"
#include "codecext.c"
- 8. Changes of files (see strings with "-> "):

== codecext.c ==  (only for sqlite-3.7.10.0)

```
else
{
    // Rollback, rekey failed
```

->        //sqlite3BtreeRollback(pbt, SQLITE_ERROR);
->        sqlite3BtreeRollback(pbt);

```
    // go back to read key
    if (HasReadKey(pCodec))
        SetWriteIsRead(pCodec);
    else //Database wasn't encrypted to start with
        sqlite3PagerSetCodec(pPager, NULL, NULL, NULL, NULL); 
}

return rc;
```

}

== codec.cpp ==

#include "codec.h"
-> //#include <botan/init.h>
-> #include <botan_all.h>

Codec::Codec(void *db)

== codec.h ==

#include <string>
#include <memory>
-> //#include <botan/botan.h>
-> //#include <botan/loadstor.h>
-> #include <botan_all.h>

using namespace std;
using namespace Botan;
- 9. Run 

msys\1.0\msys.bat
- 10. In msys-shell run commands:

cd ./Botan-1.10.6
./configure.py --gen-amalgamation
g++ -c botan_all.cpp -o botan.o
gcc -c sqlite3mod.c -o botansqlite3.o
g++ -I. -c codec.cpp -o codec.o
ar rcs libbotansqlite3.a botan.o botansqlite3.o codec.o
g++ -I. test_sqlite.cpp -o test_sqlite.exe -L. -lbotansqlite3
test_sqlite.exe
- 11. If you see "All Seems Good" then All Seems Good :)
